Fix automatic indentation in case of certain character literals

iamFIREcracker opened the pull request:
Currently vlime's automatic indentation would break if a `#\(` is found
anywhere on a buffer.

Expected behavior:

    (defun open-parenthesis-p (ch)
      (char= ch #\())
    _ ; <-- cursor position

Actual behavior:

    (defun open-parenthesis-p (ch)
      (char= ch #\())
      _ ; <-- cursor position

The fix is to change the call to `searchpairpos()` inside
`vlime#ui#ParseOuterOperators()` to skip over specific syntax types
(e.g. special, and symbol)

iamFIREcracker committed:
Fix automatic indentation in case of certain character literals

Currently vlime's automatic indentation would break if a `#\(` is found
anywhere on a buffer.

Expected behavior:

    (defun open-parenthesis-p (ch)
      (char= ch #\())
    _ ; <-- cursor position

Actual behavior:

    (defun open-parenthesis-p (ch)
      (char= ch #\())
      _ ; <-- cursor position

The fix is to change the call to `searchpairpos()` inside
`vlime#ui#ParseOuterOperators()` to skip over specific syntax types
(e.g. special, and symbol)

phmarek commented on vim/autoload/vlime/ui.vim:
```diff
@@ -583,7 +587,7 @@ function! vlime#ui#ParseOuterOperators(max_count)
     try
         while len(stack) < a:max_count
-            let [p_line, p_col] = searchpairpos('(', '', ')', 'bnW')
+            let [p_line, p_col] = searchpairpos('(', '', ')', 'bnW', s:skipped_regions)
```
Shouldn't that be `s_skip` here, as in the other `searchpairpos` calls?

There are some more `searchpairpos` uses; shouldn't they all use this (or a similar) skip argument? Perhaps we should create one or two helper functions?

phmarek reviewed (commented):
Hi,

thanks for your help! I'm not sure about one detail, though.

iamFIREcracker committed:
Revert changes to code extracted from matchparen.vim

iamFIREcracker committed:
Add `searchpairpos()` wrapper to search pairs of parentheses

iamFIREcracker committed:
Change `CurTopExprPos` to use `SearchParenPos`

Given the following buffer:

    (defun hello-world()
      (format t "Hello, world!"))

    (defun open-parenthesis-p (ch)
      "Some documentation containing unbalanced ("
      (char= ch #\())

    (defun close-parenthesis-p (ch)
      "Some documentation containing unbalanced )"
      (char= ch #\)))

Placing the cursor onto the `#\(` character literal, and running `echo
vlime#ui#CurTopExprPos('begin')` correctly logs: `[4, 1]`; however, if
we pass in `'end'` we get `[10, 17]` instead of the expected `[6, 17]`.
Similarly, if we place the cursor onto the `#\)` character literal and
run `echo vlime#ui#CurTopExprPos('end')` we get: `[10, 17]`, but if we
use `'begin'` we get `[4, 1]` instead of `[8, 1]`.  Switching to our
`searchpairpos()` wrapper seems to be fixing the problem.

iamFIREcracker committed:
Change `CurOperator` to use `SearchParenPos`

Given the following buffer:

    (defun hello-world()
      (format t "Hello, world!"))

    (defun open-parenthesis-p (ch)
      "Some documentation containing unbalanced ("
      (char= ch #\())

    (defun close-parenthesis-p (ch)
      "Some documentation containing unbalanced )"
      (char= ch #\)))

Placing the cursor onto the `#\(` character literal, and running `echo
vlime#ui#CurOperator()` returns an empty string, and interestingly
enough adding `skip` to `searchpairpos()` does not seem to make much
difference.  Why? Because the function calls `vlime#ui#CurExpr()` which
returns `()` instead of the expected `(char= ch #\())`.  Why is
`vlime#ui#CurExpr()` returning `()`? Because we are re-using the same
logic of matchparen.vim, which intentionally does _not_ try to skip over
certain syntax regions if the position the was in at the time the search
operation was issued happened to be within one of those syntax regions
already (i.e. it will highlight parentheses inside comments).

In any case, since we have to deal with incomplete expressions anyway,
I am proposing that we stop calling `vlime#ui#CurExpr()` here, and
simply use the `searchpairpos()` wrapper instead.

iamFIREcracker committed:
Change `SurroundingOperator` to use `SearchParenPos`

Given the following buffer:

    (defun hello-world()
      (format t "Hello, world!"))

    (defun open-parenthesis-p (ch)
      "Some documentation containing unbalanced ("
      (char= ch #\())

    (defun close-parenthesis-p (ch)
      "Some documentation containing unbalanced )"
      (char= ch #\)))

Placing the cursor onto the closing parenthesis right after the `#\(`
character literal and calling this function outputs an empty string
instead of `char=`.  The same happens if we place the cursor at the
closing parenthesis right after the `#\)` character literal of the
second function definition.  Using our `searchpairpos()` wrapper here,
seems to be fixing the problem.

iamFIREcracker committed:
Change `ParseOuterOperators()` to use `SearchParenPos`

Given the following buffer:

    (defun hello-world()
      (format t "Hello, world!"))

    (defun open-parenthesis-p (ch)
      "Some documentation containing unbalanced ("
      (char= ch #\())

    (defun close-parenthesis-p (ch)
      "Some documentation containing unbalanced )"
      (char= ch #\)))

Placing the cursor onto the closing parenthesis right after the `#\(`
character literal and calling `vlime#ui#ParseOuterOperators(3)` returns
`[['', -1, [6, 15]], ['char=', 2, [6, 3]], ['', 1, [5, 45]]]` instead of
the expected `[['char=', 3, [6, 3]], ['defun', 7, [4, 1]]]`.  Again,
using our `searchpairpos()` wrapper fixes the problem.

iamFIREcracker committed:
Change `CurArgPos` to use `SearchParenPos`

Given the following buffer:

    (defun open-parenthesis-yoda-p (ch)
      "Some documentation containing unbalanced ("
      (char= #\( ch))

    (defun close-parenthesis-yoda-p (ch)
      "Some documentation containing unbalanced )"
      (char= #\) ch))

Placing the cursor onto the `ch` symbol right after the `#\(` character
literal and calling `vlime#ui#CurArgPos()` outputs `0` instead of the
`2`.  Similarly, doing the same from the `ch` symbol right after the
`#\)` character literal outputs `5` instead of `2`.  To fix this, not
only we had to use our `searchpairpos()` wrapper, but also change the
_parser_ to try to deal with character literals.

iamFIREcracker commented:
> There are some more searchpairpos uses; shouldn't they all use this (or a similar) skip argument? Perhaps we should create one or two helper functions?

These are the functions which make use of `searchpairpos()`:

- `vlime#ui#CurTopExprPos()`
- `vlime#ui#CurOperator()`
- `vlime#ui#SurroundingOperator()`
- `vlime#ui#ParseOuterOperators(3)`
- `vlime#ui#CurArgPos()`

Let's load the following into a buffer and see how each of the above behave:

    (defun hello-world()
      (format t "Hello, world!"))

    (defun open-parenthesis-p (ch)
      "Some documentation containing unbalanced ("
      (char= ch #\())

    (defun close-parenthesis-p (ch)
      "Some documentation containing unbalanced )"
      (char= ch #\)))

Placing the cursor onto the `#\(` character literal, and running `echo vlime#ui#CurTopExprPos('begin')` correctly logs: `[4, 1]`; however, if we pass in `'end'` we get `[10, 17]` instead of the expected `[6, 17]`.  Similarly, if we place the cursor onto the `#\)` character literal and run `echo vlime#ui#CurTopExprPos('end')` we get: `[10, 17]`, but if we use `'begin'` we get `[4, 1]` instead of `[8, 1]`.  Switching to our `searchpairpos()` wrapper seems to be fixing the problem (see 6fe69f3).

Placing the cursor onto the `#\(` character literal, and running `echo vlime#ui#CurOperator()` returns an empty string, and interestingly enough adding `skip` to `searchpairpos()` does not seem to make much difference.  Why? Because the function calls `vlime#ui#CurExpr()` which returns `()` instead of the expected `(char= ch #\())`.  Why is `vlime#ui#CurExpr()` returning `()`? Because we are re-using the same logic of matchparen.vim, which intentionally does _not_ try to skip over certain syntax regions if the position the was in at the time the search operation was issued happened to be within one of those syntax regions already (i.e. it will highlight parentheses inside comments).

Question: is this the expected behavior, or should `vlime#ui#CurExpr()` return the actual surrounding expression instead?

In any case, since we have to deal with incomplete expressions anyway, I am proposing that we stop calling `vlime#ui#CurExpr()` here, and simply use the `searchpairpos()` wrapper instead (see debdd92).

Off to `vlime#ui#SurroundingOperator()`; placing the cursor onto the closing parenthesis right after the `#\(` character literal and calling this function outputs an empty string instead of `char=`.  The same happens if we place the cursor at the closing parenthesis right after the `#\)` character literal of the second function definition.  Using our `searchpairpos()` wrapper here, seems to be fixing the problem (see dac541a).

The next one is `vlime#ui#ParseOuterOperators()`; placing the cursor onto the closing parenthesis right after the `#\(` character literal and calling `vlime#ui#ParseOuterOperators(3)` returns `[['', -1, [6, 15]], ['char=', 2, [6, 3]], ['', 1, [5, 45]]]` instead of the expected `[['char=', 3, [6, 3]], ['defun', 7, [4, 1]]]`.  Again, using our `searchpairpos()` wrapper fixes the problem
 (see f8a8777).

Lastly, for `vlime#ui#CurArgPos()`, let's use these function definitions instead:

    (defun open-parenthesis-yoda-p (ch)
      "Some documentation containing unbalanced ("
      (char= #\( ch))

    (defun close-parenthesis-yoda-p (ch)
      "Some documentation containing unbalanced )"
      (char= #\) ch))

Placing the cursor onto the `ch` symbol right after the `#\(` character literal and calling `vlime#ui#CurArgPos()` outputs `0` instead of the `2`.  Similarly, doing the same from the `ch` symbol right after the `#\)` character literal outputs `5` instead of `2`.  To fix this, not only we had to use our `searchpairpos()` wrapper, but also change the _parser_ to try to deal with character literals (see 5ea8f32).

phmarek commented:
Had that active for some time and didn't notice any troubles, so pushed to master.

Thanks!
